src/Prefetch.cpp

```diff
@@ -0,0 +1,348 @@
+        Expr var = Variable::make(Int(32), op->name);
+        Interval prein(var, var);
+        scope.push(op->name, prein);
+        debug(dbg_prefetch5) << "For scope.push(" << op->name << ", (" << var << "))\n";
+
+        debug(dbg_prefetch4) << "For: " << op->name << " " << func_name << " " << ivar_name << "\n";
+        if (prefetches.empty()) {
+            debug(dbg_prefetch4) << " No prefetch directives in schedule\n";
+        } else {
+            debug(dbg_prefetch4) << " Found prefetch directive(s) in schedule\n";
+        }
+
+        body = mutate(body);
+
+        for (const Prefetch &p : prefetches) {
+            debug(dbg_prefetch4) << " Check prefetch ivar: " << p.var << " == " << ivar_name << "\n";
+            if (p.var == ivar_name) {
+                debug(dbg_prefetch4) << " Found directive matching " << ivar_name << "\n";
+                debug(dbg_prefetch1) << " " << func_name
+                                     << " prefetch(" << ivar_name << ", " << p.offset << ")\n";
+
+                // Add loop variable + prefetch offset to interval scope for box computation
+                Expr var = Variable::make(Int(32), op->name);
+                Interval prein(var + p.offset, var + p.offset);
+                scope.push(op->name, prein);
+                debug(dbg_prefetch5) << " For scope.push(" << op->name << ", ("
+                                     << var << " + " << p.offset << ", "
+                                     << var << " + " << p.offset << "))\n";
+
+                map<string, Box> boxes;
+                boxes = boxes_required(body, scope);
+                // TODO: Opt: prefetch the difference from previous iteration
+                //            to the requested iteration (2 calls to boxes_required)
+
+                vector<pair<string, Expr>> plets;  // Prefetch let assignments
+                vector<Stmt> pstmts;               // Prefetch stmt sequence
+                debug(dbg_prefetch3) << "  boxes required:\n";
+                for (auto &b : boxes) {
+                    const string &varname = b.first;
+                    Box &box = b.second;
+                    prefetch_box(varname, box, plets, pstmts);
+                }
+
+                inject_prefetch_stmts(op, p, plets, pstmts, body);
+
+                debug(dbg_prefetch5) << " For scope.pop(" << op->name << ")\n";
+                scope.pop(op->name);
+            }
+        }
+
+        debug(dbg_prefetch5) << "For scope.pop(" << op->name << ")\n";
+        scope.pop(op->name);
+
+        stmt = For::make(op->name, op->min, op->extent, op->for_type, op->device_api, body);
+        debug(dbg_prefetch4) << "EndFor: " << op->name << "\n";
+    }
+
+};
+
+Stmt inject_prefetch(Stmt s, const std::map<std::string, Function> &env)
+{
+    int dbg1 = 1, dbg2 = 2, dbg3 = 3, dbg4 = 4, dbg5 = 5;
+
+    size_t read;
+    std::string lvl = get_env_variable("HL_DEBUG_PREFETCH", read);
```

Comment:
I don't think there should be a separate debug level for prefetching. Then you can get rid of all of these debug level variables. As I said elsewhere though, I think most of the debugging code here should be removed.